build(test-driver-definitions): Node16 and type-only

DLehenbauer opened the pull request:
- Make test-driver-definitions emit types only and set package exports
- In test-driver-definitions use Node16 resolution for imports
- Restore test-version-utils compatOptions to TypeScript to build for both ESM and CJS which eliminates CJS imports in otherwise ESM context (and dual CJS+ESM imports of common modules). 

[AB#7400](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/7400)

DLehenbauer committed:
build(test-driver-definitions): use fluid-tsc

DLehenbauer committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test-driver-definitions

DLehenbauer committed:
Merge branch 'main' into test-driver-definitions

DLehenbauer committed:
Merge branch 'test-driver-definitions' of https://github.com/microsoft/FluidFramework into test-driver-definitions

DLehenbauer committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test-driver-definitions

DLehenbauer committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test-driver-definitions

DLehenbauer committed:
Merge branch 'main' into test-driver-definitions

sonalideshpandemsft committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test-driver-definitions

sonalideshpandemsft committed:
rm tsconfig.esnext

sonalideshpandemsft committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test-driver-definitions

jason-ha committed:
build(test-driver-definitions): make package type-only
+ emit declarations only
+ clean up extra tsconfig file
+ remove structured exports (internal only)

jason-ha committed:
build(test-driver-definitions): remove unused API rollup
generation and config files

sonalideshpandemsft committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test-driver-definitions

sonalideshpandemsft committed:
format biome

sonalideshpandemsft committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test-driver-definitions

sonalideshpandemsft committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test-driver-definitions

sonalideshpandemsft committed:
workaround

sonalideshpandemsft committed:
add comments

sonalideshpandemsft committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test-driver-definitions

sonalideshpandemsft committed:
format prettier

Abe27342 commented:
PR description/title could use some more clarity (maybe by removing unnecessary information :))

tylerbutler commented on packages/test/test-version-utils/compatOptions.d.cts:
```diff
@@ -3,7 +3,17 @@
  */
 
-import { RouterliciousEndpoint, TestDriverTypes } from "@fluidframework/test-driver-definitions";
+/**
+ * The values of RouterliciousEndpoint correspond to those defined in the interfaces file of the test-driver-definitions package,
+ * {@link @fluidframework/test-driver-definitions#RouterliciousEndpoint}
+ */
+type RouterliciousEndpoint = "frs" | "frsCanary" | "r11s" | "docker";
+
+/**
+ * The values of TestDriverTypes correspond to those defined in the interfaces file of the test-driver-definitions package,
+ * {@link @fluidframework/test-driver-definitions#TestDriverTypes}
+ */
+type TestDriverTypes = "tinylicious" | "t9s" | "routerlicious" | "r11s" | "odsp" | "local";
```
I assume if the values in test-driver-defs change, then these values also need to be updated. I suggest adding comments that describe why these values are duplicated from the ones in test-driver-defs. Maybe just add the details from the PR description as a comment so people will see it in code.

jason-ha committed:
fix(test-version-utils): restore compatOptions to typescript
partial revert of 29513d7a80

ESM build now only imports other ESM verions of other FF packages avoiding dual CJS and ESM imports.
